### PR TITLE
curveadm: add create fs before umount

### PR DESCRIPTION
**Pull request by Cyber-SiKu**
fuse-client rm create-fs in init,
so add create-fs in mount

### CODE REVIEW

File: internal/task/task/fs/mount.go [hunk 1]
```diff
@@ -153,11 +182,28 @@ func NewMountFSTask(curvradm *cli.CurveAdm, cc *client.ClientConfig) (*task.Task
...
+		ExecInLocal:       true,
+	})
 	t.AddStep(&step.StartContainer{
 		ContainerId:  &containerId,
 		ExecWithSudo: false,
 		ExecInLocal:  true,
 	})
+	t.AddStep(&step.ContainerExec{ // create fs
+		ContainerId:	&containerId,
+		Command:		fmt.Sprintf("curvefs_tool create-fs -fsName=%s -rpcTimeoutMs=%d", mountFSName, RPC_TIMEOUT_MS),
```
Comment (Wine93) [inline internal/task/task/fs/mount.go:203]: typo

File: internal/task/task/fs/mount.go [hunk 2]
```diff
@@ -153,11 +182,28 @@ func NewMountFSTask(curvradm *cli.CurveAdm, cc *client.ClientConfig) (*task.Task
...
+		Mutate:            newToolsMutate(cc, CLIENT_CONFIG_DELIMITER),
+		ExecWithSudo:      false,
+		ExecInLocal:       true,
+	})
 	t.AddStep(&step.StartContainer{
 		ContainerId:  &containerId,
 		ExecWithSudo: false,
 		ExecInLocal:  true,
 	})
+	t.AddStep(&step.ContainerExec{ // create fs
```
Comment (Wine93) [inline internal/task/task/fs/mount.go:201]: should we execute `create fs` after mount done, is it works?

File: internal/task/task/fs/mount.go [hunk 3]
```diff
@@ -39,6 +39,9 @@ const (
 	FORMAT_MOUNT_OPTION = "type=bind,source=%s,target=%s,bind-propagation=rshared"
 
 	CLIENT_CONFIG_DELIMITER = "="
+
+	LAYOUT_CURVEFS_TOOLS_CONFIG_SYSTEM_PATH = "/etc/curvefs/tools.conf"
```
Comment (Wine93) [inline internal/task/task/fs/mount.go:43]: It's better to define a module level funtion in topology module to reduce redefine, like `GetProjectLayout`.
[see also](https://github.com/opencurve/curveadm/blob/d7e25679c8452b01dba4e00e73caa60896aa1421/internal/configure/topology/dc_get.go#L238)

File: internal/task/task/fs/mount.go [hunk 4]
```diff
@@ -153,11 +182,28 @@ func NewMountFSTask(curvradm *cli.CurveAdm, cc *client.ClientConfig) (*task.Task
 		ExecWithSudo:      false,
 		ExecInLocal:       true,
 	})
+
+	t.AddStep(&step.SyncFile{ // sync tools config
+		ContainerSrcId:    &containerId,
+		ContainerSrcPath:  fmt.Sprintf("%s/conf/tools.conf", root),
+		ContainerDestId:   &containerId,
+		ContainerDestPath: LAYOUT_CURVEFS_TOOLS_CONFIG_SYSTEM_PATH,
```
Comment (Wine93) [inline internal/task/task/fs/mount.go:190]: ditto

File: internal/task/task/fs/mount.go [hunk 5]
```diff
@@ -110,6 +113,32 @@ func newMutate(cc *client.ClientConfig, delimiter string) step.Mutate {
 	}
 }
 
+func newToolsMutate(cc *client.ClientConfig, delimiter string) step.Mutate {
```
Comment (Wine93) [inline internal/task/task/fs/mount.go:116]: Maybe the logic can more clearly:
```
clientConfig := cc.GetServiceConfig()
tools2client := map[string]string {
    "mdsAddr" : "mdsOpt.rpcRetryOpt.addrs"
}
return func(in, key, value string) (out string, err error) {
    if len(key) == 0 || tools2client[key] == 0 {
        out = in
        return
    }
    key = tools2client[key]
    out = fmt.Sprintf("%s%s%s", key, delimiter, clientConfig[key])
    return
}
```

File: internal/task/task/fs/mount.go (hunk 2 above)
Comment (Cyber-SiKu) [reply internal/task/task/fs/mount.go:201]: > should we execute `create fs` after mount done and is it works?

In this case, any process that accesses the mounted directory will get stuck. If the creation fails, should not continue.

File: internal/task/task/fs/mount.go (hunk 5 above)
Comment (Cyber-SiKu) [reply internal/task/task/fs/mount.go:116]: > Maybe the logic can more clearly:
> 
> ```
> clientConfig := cc.GetServiceConfig()
> tools2client := map[string]string {
>     "mdsAddr" : "mdsOpt.rpcRetryOpt.addrs"
> }
> return func(in, key, value string) (out string, err error) {
>     if len(key) == 0 || tools2client[key] == 0 {
>         out = in
>         return
>     }
>     key = tools2client[key]
>     out = fmt.Sprintf("%s%s%s", key, delimiter, clientConfig[key])
>     return
> }
> ```

fix

File: internal/configure/topology/dc_get.go [hunk 6]
```diff
@@ -215,3 +215,8 @@ func (dc *DeployConfig) GetProjectLayout() Layout {
 		CoreSystemDir: LAYOUT_CORE_SYSTEM_DIR,
 	}
 }
+
+func GetProjectLayout(kind string) Layout {
```
Comment (Wine93) [inline internal/configure/topology/dc_get.go:219]: 👍

File: internal/task/scripts/createfs.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+function main() {
+    g_fsname=$g_fsname$1
+
+    $g_curvefs_tool $g_curvefs_tool_operator $g_fsname $g_rpc_timeout_ms 
+}
+
+main "$@"
+
+if [ $? -eq 0 ]; then
+    $g_entrypoint "$@"
```
Comment (Wine93) [inline internal/task/scripts/createfs.go:45]: Why add these after `main` function? In general, `main` is the entrypoint of program.

File: internal/task/task/fs/mount.go [hunk 8]
```diff
@@ -129,7 +162,7 @@ func NewMountFSTask(curvradm *cli.CurveAdm, cc *client.ClientConfig) (*task.Task
 	t.AddStep(&step.CreateContainer{
 		Image:             cc.GetContainerImage(),
 		Envs:              []string{"LD_PRELOAD=/usr/local/lib/libjemalloc.so"},
-		Command:           getMountCommand(cc, mountFSName),
+		Command:           getPreMountCommand(mountFSName) + getMountCommand(cc, mountFSName),
```
Comment (Wine93) [inline internal/task/task/fs/mount.go:165]: Maybe join these command into one is better.

File: internal/task/scripts/createfs.go (hunk 7 above)
Comment (Cyber-SiKu) [reply internal/task/scripts/createfs.go:45]: > Why add these after `main` function? In general, `main` function is the entrypoint of program.

fix

File: internal/task/task/fs/mount.go (hunk 8 above)
Comment (Cyber-SiKu) [reply internal/task/task/fs/mount.go]: > Maybe join these command into one is better.

fix

**Pull request merged**